core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -489,15 +498,16 @@ public void relocated(String reason) throws IllegalIndexShardStateException, Int
                     if (state != IndexShardState.STARTED) {
                         throw new IndexShardNotStartedException(shardId, state);
                     }
-                    // if the master cancelled the recovery, the target will be removed
-                    // and the recovery will stopped.
-                    // However, it is still possible that we concurrently end up here
-                    // and therefore have to protect we don't mark the shard as relocated when
-                    // its shard routing says otherwise.
+                    /*
+                     * If the master cancelled recovery, the target will be removed and the recovery will be cancelled. However, it is still
+                     * possible that we concurrently end up here and therefore have to protect that we do not mark the shard as relocated
+                     * when its shard routing says otherwise.
+                     */
                     if (shardRouting.relocating() == false) {
                         throw new IllegalIndexShardStateException(shardId, IndexShardState.STARTED,
                             ": shard is no longer relocating " + shardRouting);
                     }
+                    onBlocked.run();
```

Comment:
This is a good observation, and I agree we should not do this. It can implicitly block the cluster state update thread on this network operation. I pushed ccde79826cf09277235d6ef78ce7870beb6273d3.